crates/wasmtime/src/module/serialization.rs

```diff
@@ -0,0 +1,715 @@
+            let type_tables_idx = *type_tables_pushed.entry(ptr as usize).or_insert_with(|| {
+                type_tables.push(MyCow::Borrowed(module.types()));
+                type_tables.len() - 1
+            });
+            SerializedModuleData {
+                artifacts: module
+                    .inner
+                    .artifact_upvars
+                    .iter()
+                    .map(|i| MyCow::Borrowed(i.compilation_artifacts()))
+                    .chain(Some(MyCow::Borrowed(
+                        module.compiled_module().compilation_artifacts(),
+                    )))
+                    .collect(),
+                modules: module
+                    .inner
+                    .module_upvars
+                    .iter()
+                    .map(|i| module_data(i, type_tables_pushed, type_tables))
+                    .collect(),
+                type_tables: type_tables_idx,
+            }
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SerializedModule<'a> {
+    version: String,
+    target: String,
+    flags_hash: u64,
+    // Record the opt level as it is the most common Cranelift flag users might change
+    opt_level: OptLevel,
+    isa_flags: Vec<String>,
+    strategy: CompilationStrategy,
+    tunables: Tunables,
+    features: WasmFeatures,
+    data: SerializedModuleData<'a>,
+    tables: Vec<MyCow<'a, TypeTables>>,
+}
+
+impl<'a> SerializedModule<'a> {
+    pub fn new(module: &'a Module) -> Self {
+        let (data, tables) = SerializedModuleData::new(module);
+        Self::with_data(module.engine().compiler(), data, tables)
+    }
+
+    pub fn from_artifacts(
+        compiler: &Compiler,
+        artifacts: &'a Vec<CompilationArtifacts>,
+        types: &'a TypeTables,
+    ) -> Self {
+        Self::with_data(
+            compiler,
+            SerializedModuleData {
+                artifacts: artifacts.iter().map(MyCow::Borrowed).collect(),
+                modules: Vec::new(),
+                type_tables: 0,
+            },
+            vec![MyCow::Borrowed(types)],
+        )
+    }
+
+    fn with_data(
+        compiler: &Compiler,
+        data: SerializedModuleData<'a>,
+        tables: Vec<MyCow<'a, TypeTables>>,
+    ) -> Self {
+        let isa = compiler.isa();
+
+        Self {
+            version: env!("CARGO_PKG_VERSION").to_string(),
+            target: isa.triple().to_string(),
+            opt_level: isa.flags().opt_level().into(),
+            flags_hash: Self::simple_hash(isa.flags()),
+            isa_flags: isa.enabled_isa_flags(),
+            strategy: compiler.strategy(),
+            tunables: compiler.tunables().clone(),
+            features: compiler.features().into(),
+            data,
+            tables,
+        }
+    }
+
+    pub fn into_module(self, engine: &Engine) -> Result<Module> {
+        let compiler = engine.compiler();
+        let isa = compiler.isa();
+
+        self.check_version()?;
```

Comment:
With these various `check_*` methods we can probably remove the `Hash` implementation for `Compiler` since IIRC it's only really used for this?